src/network/endpoint.ml

```diff
@@ -1,13 +1,56 @@
 open Deku_consensus
+open Deku_concepts
+
+type 'a post = Post of 'a
+type get = Get
+
+type _ endpoint =
+  | Consensus : External_message.packed post endpoint
+  | Get_block_by_level : Level.t -> get endpoint
 
-type _ endpoint = Consensus : External_message.packed endpoint
 type 'a t = 'a endpoint
 type ex = Ex : _ endpoint -> ex
 
 let consensus = Consensus
 
-let of_string path =
-  match path with "/consensus" -> Some (Ex Consensus) | _ -> None
+(* Just a helper to catch the exception of yojson, TODO: should it be defined here ?? *)
+let of_yojson t_of_yojson yojson =
+  try t_of_yojson yojson |> Option.some with _ -> None
+
+module Consensus_route = struct
+  let path () = Routes.(s "consensus" /? nil)
+  let parser () = Routes.(path () @--> Ok (Ex Consensus))
+end
+
+module Get_block_by_level_route = struct
+  let path () = Routes.(s "chain" / s "blocks" / str /? nil)
+
+  let parser () =
+    Routes.(
+      path () @--> fun level ->
+      `String level
+      |> of_yojson Level.t_of_yojson
+      |> Option.to_result ~none:(Internal_error.invalid_level level)
+      |> Result.map (fun level -> Ex (Get_block_by_level level)))
+end
+
+let routes =
```

Comment:
This seems way better than the previous `split`!
What about the performance? Is there any benchmark on this tool?

Reply:
Can't find the benchmarks currently, but it's FAST. [Ahrefs seems to use it](https://discuss.ocaml.org/t/your-production-web-stack-in-2020/6691/8) which should be a good indicator of the quality.